aten/src/ATen/native/sparse/SparseFactories.h

```diff
@@ -0,0 +1,139 @@
+#pragma once
+#include <ATen/TensorIndexing.h>
+#include <ATen/TensorIterator.h>
+#include <ATen/core/ATen_fwd.h>
+#include <ATen/core/Tensor.h>
+#include <ATen/native/TensorFactories.h>
+#include <c10/core/Scalar.h>
+#include <c10/util/ArrayRef.h>
+#include <c10/util/Exception.h>
+
+#ifndef AT_PER_OPERATOR_HEADERS
+#include <ATen/Functions.h>
+#include <ATen/NativeFunctions.h>
+#else
+#include <ATen/ops/sparse_coo_tensor.h>
+#endif
+
+namespace at {
+namespace native {
+namespace impl {
+
+template <typename kernel_func_t>
+Tensor spdiags_impl(
+    const Tensor& diagonals,
+    const Tensor& offsets,
+    IntArrayRef shape,
+    c10::optional<Layout> layout,
+    const kernel_func_t& kernel_func) {
+  auto diagonals_2d = diagonals.dim() == 1 ? diagonals.unsqueeze(0) : diagonals;
+  TORCH_CHECK(diagonals_2d.dim() == 2, "Diagonals must be vector or matrix");
+  TORCH_CHECK(shape.size() == 2, "Output shape must be 2d");
+  auto offsets_1d = offsets.dim() == 0 ? offsets.unsqueeze(0) : offsets;
+  TORCH_CHECK(offsets_1d.dim() == 1, "Offsets must be scalar or vector");
+  TORCH_CHECK(
+      diagonals_2d.size(0) == offsets_1d.size(0),
+      "Number of diagonals (",
+      diagonals_2d.size(0),
+      ") does not match the number of offsets (",
+      offsets_1d.size(0),
+      ")");
+  if (layout) {
+    TORCH_CHECK(
+        (*layout == Layout::Sparse) || (*layout == Layout::SparseCsc) ||
+            (*layout == Layout::SparseCsr),
+        "Only output layouts (",
+        Layout::Sparse,
+        ", ",
+        Layout::SparseCsc,
+        ", and ",
+        Layout::SparseCsr,
+        ") are supported");
+  }
+  TORCH_CHECK(
+      offsets_1d.scalar_type() == at::kLong,
+      "spdiags(): Expected a LongTensor of offsets but got ",
+      offsets_1d.scalar_type());
+
+  TORCH_CHECK(
+      offsets_1d.unsqueeze(0)
+          .permute({1, 0})
+          .eq(offsets_1d)
+          .sum(-1)
+          .equal(at::ones_like(offsets_1d)),
+      "spdiags(): Offset tensor contains duplicate values");
+
+  // We are going to introduce some temporaries to compute nnz_per_diag and do
+  // the final check at onece, we induce a block scope to allow these
+  // intermediates to be cleaned up before we continue
+  auto nnz_per_diag = at::empty_like(offsets_1d);
+  {
+    auto n_col_out = at::scalar_tensor(shape[1], offsets_1d.options());
```

Comment:
I think we could compute `nnz_per_diag` with a single TensorIterator loop over diagonals with all the checks. The only problem is that this will require device-specific error handling with `CUDA_KERNEL_ASSERT`. Not sure this operation is that performance critical, but we could add a comment about that, however.